### PR TITLE
SCOPE_IDENTITY should return identity value in correct scope.

### PR DESCRIPTION
**Pull request by lejaokri**
IDENTITY and SCOPE_IDENTITY are the same today because they have the same implementation. This is wrong. In order to fix this, we introduce a stack of identity value which should contain the correct identity value for a particular scope and return that when a user calls select scope_identity().

On the other hand, select @@identity remains the same.

The stack implementation is similar to the GucStack.

Task: [BABEL-3907]
Signed-off-by: Kristian Lejao <klejao@amazon.com>

### Test Scenarios Covered ###
-- Test scenarios
(1) Fixed existing SCOPE_IDENTITY testcases: (Test-Identity and identity_function)
(2) Introduced new testcase Test-scope-identity to test out exact scenario in BABEL-3907
    and additional coverage

Overall, the three testcases cover the following:

(1) Multiple tables with triggers
(2) Stored procedures calling INSERT statement with TRIGGER
(3) Stored procedures calling another SP calling INSERT with TRIGGER
(4) SCOPE_IDENTITY inside a View
(5) SCOPE_IDENTITY inside a Function
(6) SCOPE_IDENTITY inside sp_executesql
(7) MVU testcase before 14_6/15_2 and newer

Outputs were validated against SQL Server

### Check List
- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: contrib/babelfishpg_tsql/sql/sys_functions.sql [hunk 1]
```diff
@@ -236,7 +236,7 @@ LANGUAGE C STABLE;
 CREATE OR REPLACE FUNCTION sys.scope_identity()
 RETURNS numeric(38,0) AS
 $BODY$
-	SELECT sys.babelfish_get_last_identity_numeric()::numeric(38,0);
+	SELECT sys.babelfish_get_scope_identity_numeric()::numeric(38,0);
```
Comment (macwoody) [inline contrib/babelfishpg_tsql/sql/sys_functions.sql]: I think we don't need the intermediate functions. 
Just letting scope_idenity be implemented with the C function returning an numeric(38,0) would be enough

File: test/python/expected/upgrade_validation/expected_dependency.out [hunk 2]
```diff
@@ -135,6 +135,8 @@ Function sys.babelfish_get_login_default_db(text)
 Function sys.babelfish_get_microsecs_from_fractsecs(text,numeric)
 Function sys.babelfish_get_monthnum_by_name(text,jsonb)
 Function sys.babelfish_get_name_delimiter_pos(text)
+Function sys.babelfish_get_scope_identity()
+Function sys.babelfish_get_scope_identity_numeric()
```
Comment (macwoody) [inline test/python/expected/upgrade_validation/expected_dependency.out]: This means we don't have a stored procedure / view directly depending on them. 
We could avoid them by not having them.

File: test/JDBC/expected/Test-Identity-before-14_7-or-15_2-vu-verify.out [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+125
+~~END~~
+
+~~START~~
+numeric
+125
+~~END~~
+
+
+SELECT * FROM test_identity_vu_prepare_t2
```
Comment (macwoody) [inline test/JDBC/expected/Test-Identity-before-14_7-or-15_2-vu-verify.out]: I think we need order by here

File: test/JDBC/expected/Test-Identity-before-14_7-or-15_2-vu-prepare.out [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+CREATE FUNCTION test_identity_vu_prepare_func4()
+RETURNS TINYINT
+AS
+BEGIN
+RETURN IDENT_CURRENT('test_identity_vu_prepare_t5')
+END
+GO
+
+--test for OUTPUT Clause
+CREATE TABLE test_identity_vu_prepare_t6
```
Comment (macwoody) [inline test/JDBC/expected/Test-Identity-before-14_7-or-15_2-vu-prepare.out:206]: Do we have a case representing the reported issue in the test? 
Can we have a few more cases with more complex nesting ?  like stored procedure -> sp_executesql -> trigger

File: test/JDBC/expected/Test-Identity-before-14_7-or-15_2-vu-verify.out (hunk 3 above)
Comment (lejaokri) [reply test/JDBC/expected/Test-Identity-before-14_7-or-15_2-vu-verify.out:148]: The query is the following I dont think we need order by

```
SELECT MAX(id) as MaximumUsedIdentity FROM test_identity_vu_prepare_t1
SELECT SCOPE_IDENTITY()
SELECT @@IDENTITY
SELECT IDENT_CURRENT('test_identity_vu_prepare_t1')
GO
```

Comment (lejaokri) [reply test/JDBC/expected/Test-Identity-before-14_7-or-15_2-vu-verify.out]: done

File: test/JDBC/expected/Test-Identity-before-14_7-or-15_2-vu-prepare.out (hunk 4 above)
Comment (lejaokri) [reply test/JDBC/expected/Test-Identity-before-14_7-or-15_2-vu-prepare.out:206]: Yes I've added the scenario in Test-scope-identity.sql (new file)

I've modified Test-scope-identity to do SP->sp_executesql->trigger

**Review by macwoody — DISMISSED** [review]

**Review by macwoody — APPROVED** [review]

**Pull request merged**